Net/SMTP.php

```diff
@@ -623,23 +623,69 @@ public function starttls()
                      * and STREAM_CRYPTO_METHOD_SSLv23_CLIENT constant is
                      * inconsistent across PHP versions. */
-                    $crypto_method = STREAM_CRYPTO_METHOD_TLS_CLIENT
-                    | @STREAM_CRYPTO_METHOD_TLSv1_1_CLIENT
-                    | @STREAM_CRYPTO_METHOD_TLSv1_2_CLIENT;
+                    $crypto_method = STREAM_CRYPTO_METHOD_TLS_CLIENT;
+
+                    if (defined('STREAM_CRYPTO_METHOD_TLSv1_1_CLIENT')) {
+                        $crypto_method |= @STREAM_CRYPTO_METHOD_TLSv1_1_CLIENT;
+                    }
+
+                    if (defined('STREAM_CRYPTO_METHOD_TLSv1_2_CLIENT')) {
+                        $crypto_method |= @STREAM_CRYPTO_METHOD_TLSv1_2_CLIENT;
+                    }
+
+                    //New Crypto-Method since PHP7.4
+                    if (defined('STREAM_CRYPTO_METHOD_TLSv1_3_CLIENT')) {
+                        $crypto_method |= @STREAM_CRYPTO_METHOD_TLSv1_3_CLIENT;
+                    }
                 }
-                if (PEAR::isError($result = $this->socket->enableCrypto(true, $crypto_method))) {
+
+                for ($attempts = 1; $attempts < 100; $attempts++) {
+                    $result = $this->socket->enableCrypto(true, $crypto_method);
+                    if ($this->socket->isBlocking() !== true) {
+                        usleep(1);
+                    }
+                    if ($result !== 0) {
+                        break;
+                    }
+                }
+
+                if (PEAR::isError($result)) {
                     return $result;
                 } elseif ($result !== true) {
-                    return PEAR::raiseError('STARTTLS failed');
+                    $crypto_types_arr = array_filter(
+                        get_defined_constants(false),
+                        function($key) {
+                            return substr($key,0,21) == 'STREAM_CRYPTO_METHOD_';
+                        },
+                        ARRAY_FILTER_USE_KEY
+                    );
+
+                    $error_types_arr = array_filter(
+                        get_defined_constants(false),
+                        function($key) {
+                            return substr($key,0,2) == 'E_';
+                        },
+                        ARRAY_FILTER_USE_KEY
+                        );
+
+                    return PEAR::raiseError('STARTTLS failed ' .
+                        '{enableCrypto: ' . var_export($result, true) . '; ' .
+                        'crypto_method: ' .
+                        array_search($crypto_method, $crypto_types_arr) .
+                        ' (' . var_export($crypto_method, true) . '); ' .
+                        'attempts: ' . $attempts . '; ' .
+                        array_search(error_get_last()['type'], $error_types_arr) .
+                        ' (' . error_get_last()['type'] . '): ' .
+                        error_get_last()['message'] . '}');
```

Comment:
I think this would be more readable using [`sprintf()`](https://www.php.net/sprintf).

Reply:
I will check that and give response soon.

Reply:
@jparise I changed a few things to get better readability and backwards-compatibility to PHP5.4. Can you check my changes, please?